### PR TITLE
Issue #967 - Map DataDefinition to StorageArea on node

### PR DESCRIPTION
**Pull request by cobarzanc**
Optimize search of variables

### CODE REVIEW

File: TypeCobol/Compiler/Nodes/Node.cs [hunk 1]
```diff
@@ -643,6 +643,35 @@ private void Dump(StringBuilder str, int i)
         public bool IsInsideCopy() {
             return CodeElement != null && CodeElement.IsInsideCopy();
         }
+
+        public IDictionary<StorageArea, Tuple<string,DataDefinition>> StorageAreaReadsDataDefinition { get; internal set; }
+        public IDictionary<StorageArea, Tuple<string,DataDefinition>> StorageAreaWritesDataDefinition { get; internal set; }
+
+        public IEnumerable<Tuple<string,DataDefinition>> GetVariableFromNodeStorageArea(StorageArea storageArea,
```
Comment (smedilol) [inline TypeCobol/Compiler/Nodes/Node.cs]: As StorageArea is the key, why use its SymbolReference to the associated value ?

File: TypeCobol/Compiler/Diagnostics/CrossChecker.cs [hunk 2]
```diff
@@ -208,183 +208,219 @@ public static void CheckPicture(Node node, CommonDataDescriptionAndDataRedefines
...
+            }
+            if (!isReadStorageArea && node.StorageAreaWritesDataDefinition == null)
+            {
+                node.StorageAreaWritesDataDefinition = new Dictionary<StorageArea, Tuple<string,DataDefinition>>();
+            }
 
-            var isPartOfTypeDef = (node as DataDefinition) != null && ((DataDefinition) node).IsPartOfATypeDef;
-            foundQualified =
+            //search for existing daata definitinon before constructing one
+            Tuple<string,DataDefinition> searchExistingDataDefinition = node.GetVariableFromNodeStorageArea(storageArea,isReadStorageArea)?.FirstOrDefault();
```
[the remark this answers is not in the document]
Comment (cobarzanc) [reply TypeCobol/Compiler/Diagnostics/CrossChecker.cs]: We have to get information from the Node (which is stored as the string - the completeQualified Name). We have to call IndexAndFlagDataDefiniton for every found data definition, which will make context-specific action on the node.

File: TypeCobol/Compiler/Diagnostics/CrossChecker.cs [hunk 3]
```diff
@@ -489,20 +525,20 @@ private static void CheckVariable(Node node, StorageArea wname, object sent)
             DataDefinition sendingTypeDefinition = null, receivingTypeDefinition = null;
 
             if (sent == null || wname == null) return; //Both items needed
-            var wsymbol = CrossCompleteChecker.CheckVariable(node, wname);
+            var wsymbol = CrossCompleteChecker.CheckVariable(node, wname,false);
```
[the remark this answers is not in the document]
Comment (cobarzanc) [reply TypeCobol/Compiler/Diagnostics/CrossChecker.cs]: There are nodes that do not have a CodeElement.StorageReadArea or StorageWriteArea and we need to keep using CheckVariable. The method will construct/use the StorageAreaReadsDataDefinition or StorageAreaWritesDataDefinition dictionaries.

File: TypeCobol/Compiler/Diagnostics/CrossChecker.cs [hunk 4]
```diff
@@ -489,20 +525,20 @@ private static void CheckVariable(Node node, StorageArea wname, object sent)
...
             {
-                var sname = sent as QualifiedName;
                 var ssymbol = GetSymbol(node.SymbolTable, sname);
                 if (ssymbol == null) return; // sending symbol name unresolved
                 sendingTypeDefinition = ssymbol.TypeDefinition ?? GetDataDefinitionType(node.SymbolTable, ssymbol);
             }
             else if (sent is StorageArea)
             {
-                var rsymbol = CrossCompleteChecker.CheckVariable(node, (StorageArea) sent);
+                var rsymbol = CrossCompleteChecker.CheckVariable(node, (StorageArea) sent,true);
```
[the remark this answers is not in the document]
Comment (cobarzanc) [reply TypeCobol/Compiler/Diagnostics/CrossChecker.cs]: There are nodes that do not contain a CodeElement.StorageReadArea or StorageWriteArea and we need to keep using CheckVariable. The method will construct/use the StorageAreaReadsDataDefinition or StorageAreaWritesDataDefinition dictionaries.

File: TypeCobol/Compiler/Diagnostics/CrossChecker.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (smedilol) [reply TypeCobol/Compiler/Diagnostics/CrossChecker.cs]: Ok, my question was not clear, why not just get the Tuple<string,DataDefinition> using the key storageArea?

File: TypeCobol/Compiler/Diagnostics/CrossChecker.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (smedilol) [reply TypeCobol/Compiler/Diagnostics/CrossChecker.cs]: Which one? Because it sounds like a bug.

[the remark this answers is not in the document]
Comment (cobarzanc) [reply TypeCobol/Compiler/Diagnostics/CrossChecker.cs]: I found the following types of node: MOVE, SET, COMPUTE that do not have a CodeElement with StorageArea associated. 
Also, we need to call IndexAndFlagDataDefiniton to update the node context. 
The DataDefinitions found are context dependent. CheckVariable takes care of this operation and searches for a DataDefinition either in the dictionaries either in the symbolTable (and stores the found DataDefinition in the dictionary)

File: TypeCobol/Compiler/Nodes/Node.cs (hunk 1 above)
Comment (cobarzanc) [reply TypeCobol/Compiler/Nodes/Node.cs]: It was the way a DataDefinition was returned from the SymbolTable. I updated the search using directly the StorageArea key from the dictionary.

File: TypeCobol/Compiler/Diagnostics/CrossChecker.cs [hunk 5]
```diff
@@ -208,103 +228,174 @@ public static void CheckPicture(Node node, CommonDataDescriptionAndDataRedefines
...
+                node.StorageAreaReadsDataDefinition = new Dictionary<StorageArea, Tuple<string, DataDefinition>>();
+            }
+            if (!isReadStorageArea && node.StorageAreaWritesDataDefinition == null)
+            {
+                node.StorageAreaWritesDataDefinition = new Dictionary<StorageArea, Tuple<string,DataDefinition>>();
+            }
 
-            var isPartOfTypeDef = (node as DataDefinition) != null && ((DataDefinition) node).IsPartOfATypeDef;
-            foundQualified =
+            //search for existing data definitinon before constructing one
```
Comment (smedilol) [inline TypeCobol/Compiler/Diagnostics/CrossChecker.cs]: Do we need to search for existing data?
As CheckVariable is only called when the Visitor begin node, it shouldn't be.
I've created #996 to investigate this..

**Comment by smedilol** [comment]
3 comment on the code, one is already covered by issue #996.
I'll create an issue to check all our double cast in the code.

**Pull request merged**